File: AdministracionDeTienda/src/gestorAplicacion/clasesHerencia/Intervenido.java
package gestorAplicacion.clasesHerencia;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import gestorAplicacion.clasesBase.Producto;
import gestorAplicacion.clasesHerencia.Empleado.Cargo;



//Colaboradores
//Maria Alejandra Echavarria Correa


public class Intervenido extends Producto  {
	private Colecciones coleccion;
	private String color;
	private String corte;
	private double precio;
	private String estampado;

	
	public Intervenido(Tipo tipo) {
		super(tipo);
	}
	//Constructor utilizado para crear objetos de tipo intervenido a partir de objetos de tipo producto
	public Intervenido(Producto producto,Colecciones coleccion) {
		super(producto.getTipo(),producto.getCosto());
		this.coleccion=coleccion;
	}
	
	//Enumerado de colecciones para fijar las caracteristicas de los intervenidos
	public enum Colecciones{
		INDUSTRIAL(1),GEOMETRIA(2),NATURALEZA(3);

		 int id;
		 Colecciones(int i) {
			id=i;
		}
		
		public int getId() {
			return id;
		}
	}
	
	//SEGUNDA INTERACCION-ENVIO
	//Se encarga de crear los objetos de tipo intervenido a partir de los una lista de prodcutos, mediante una seleccion de operarios.
	public static ArrayList<Intervenido> intervenir(ArrayList<Producto> productos,Colecciones coleccion){
		ArrayList<Intervenido> intervenidos=new ArrayList<Intervenido> ();
		ArrayList<Empleado> operarios=Empleado.seleccionarEmpleados(Cargo.OPERARIO,productos);
		ArrayList<String> paletaDeColores=Intervenido.paletaDeColores(coleccion);
		//Verificar que se tienen operarios disponibles para realizar la intervencion
		if(!operarios.isEmpty()) {
			for(int i=0;i<productos.size();i++) {
				Intervenido intervenido=new Intervenido(productos.get(i),coleccion);
				intervenido.cortar();
				intervenido.pintar(paletaDeColores);
				intervenido.estampar();
				intervenido.fijacionDePrecio();
				intervenidos.add(intervenido);
			}
			
		}
		return intervenidos;
			
		}
	
	//Se encarga de fijar una parleta de colores dependiendo de la coleccion
	public static ArrayList<String> paletaDeColores(Colecciones coleccion) {
		ArrayList<String> paleta=new ArrayList<String>();
		List<String> colores=null;
		switch(coleccion){
		case NATURALEZA: colores= Arrays.asList("VERDE","AMARILLO","CAFE","BEIGE","OLIVO","AZUL","NEGRO");
		break;
		case GEOMETRIA: colores= Arrays.asList("AZUL","NARANJADO","ROJO","VERDE","AMARILLO","VIOLETA","NEGRO");
		break;
		case INDUSTRIAL: colores= Arrays.asList("GRIS","AZUL","VIOLETA","VERDE","PLATEADO","ROJO","NEGRO");
		break;	
		}
		paleta.addAll(colores);
		return paleta;
			
		
	}
	
	//Fija el color de los productos intervenidos escogiendo aleatoriamente un color de la paleta de colores
	public void pintar(ArrayList<String> paleta) {
		Random random = new Random();
      int indiceAleatorio = random.nextInt(paleta.size());
      String color = paleta.get(indiceAleatorio);
      this.setColor(color);
		
	}
	
	//Fija un corte segun el tipo de producto y la coleccion
	public void cortar() {
		Tipo tipo=this.getTipo();
		switch(tipo) {
		case ABRIGO:
			switch(coleccion) {
			case GEOMETRIA:setCorte("Chaqueta");
				break;
			case INDUSTRIAL:setCorte("Cazadora");
				break;
			case NATURALEZA: setCorte("Chal");
				break;
			default:
				break;
			}
			break;
		case CAMISA:
			switch(coleccion) {
			case GEOMETRIA:setCorte("Corte Cuadrado");
			break;
		    case INDUSTRIAL:setCorte("Botones");
			break;
		    case NATURALEZA: setCorte("Manga Sisa");
			break;
		default:
			break;
		}
			break;
		case PANTALON:
			switch(coleccion) {
			case GEOMETRIA:setCorte("Bermuda");
			break;
		    case INDUSTRIAL:setCorte("Traje");
			break;
		    case NATURALEZA: setCorte("Short");
			break;
		default:
			break;
		}
			break;
		
		}
		
	}
	//Fija un estampado segun el tipo de producto y la coleccion
	public void estampar() {
		Tipo tipo=this.getTipo();
		switch(tipo) {
		case ABRIGO:
			switch(coleccion) {
			case GEOMETRIA:setEstampado("Cuadrado");
				break;
			case INDUSTRIAL:setEstampado("Manchas");
				break;
			case NATURALEZA: setEstampado("Hojas");
				break;
			default:
				break;
			}
			break;
		case CAMISA:
			switch(coleccion) {
			case GEOMETRIA:setEstampado("Triangulo");
				break;
			case INDUSTRIAL:setEstampado("Edificio");
				break;
			case NATURALEZA: setEstampado("Palma");
				break;
			}
			break;
		case PANTALON:
			switch(coleccion) {
			case GEOMETRIA:setEstampado("Rombos");
				break;
			case INDUSTRIAL:setEstampado("Rayas");
				break;
			case NATURALEZA: setEstampado("Flores");
				break;
			}
			break;
		
		}
		
	}
	//Fija el precio de los objetos intervenidos a partir del costo incial del producto y la coleccion
	public void fijacionDePrecio() {
		
		switch(coleccion) {
		case NATURALEZA: this.precio=this.getCosto()+200000;
		break;
		case INDUSTRIAL: this.precio=this.getCosto()+300000;
		break;
		case GEOMETRIA: this.precio=this.getCosto()+400000;
		default: this.precio=this.getCosto();
			break;
		}
	}
	
	@Override
	//Sobreescritura para imprimir los objetos
	public String toString() {
		return "*"+this.getTipo()+" color: " + color + ", corte:" + corte + ", estampado:" + estampado+ ", precio:" + precio; 
				
	}
//Setters y getters
	public String getColor() {
		return color;
	}

	public void setColor(String color) {
		this.color = color;
	}

	public String getCorte() {
		return corte;
	}

	public void setCorte(String corte) {
		this.corte = corte;
	}

	public double getPrecio() {
		return precio;
	}

	public void setPrecio(double precio) {
		this.precio = precio;
	}

	public Colecciones getColeccion() {
		return coleccion;
	}

	public void setColeccion(Colecciones coleccion) {
		this.coleccion = coleccion;
	}

	public String getEstampado() {
		return estampado;
	}

	public void setEstampado(String estampado) {
		this.estampado = estampado;
	}


	
}

File: Python/src/gestorAplicacion/clasesHerencia/Intervenido.py
from enum import Enum
import random
from src.gestorAplicacion.clasesBase.Producto import Producto
Tipo=Producto.Tipo


    
    

class Intervenido(Producto):
    class Colecciones(Enum):
        INDUSTRIAL = 1
        GEOMETRIA = 2
        NATURALEZA = 3
    

    def __init__(self, producto,coleccion):
        super().__init__(producto.get_tipo(),producto.get_costo())
        self.coleccion = coleccion
        self.color = ""
        self.corte = ""
        self.precio = 0.0
        self.estampado = ""

    def __str__(self):

        return "*" + self.get_tipo() + " color: " + self.color + ", corte: " + self.corte + ", estampado: " + self.estampado + ", precio: " + str(self.precio)

    def pintar(self, paleta):
        indice_aleatorio = random.randint(0, len(paleta) - 1)
        self.color = paleta[indice_aleatorio]

    def cortar(self):
        tipo = self.get_tipo()
        mapeo_cortes = {
            (Tipo.ABRIGO, Intervenido.Colecciones.GEOMETRIA): "Chaqueta",
            (Tipo.ABRIGO, Intervenido.Colecciones.INDUSTRIAL): "Cazadora",
            (Tipo.ABRIGO, Intervenido.Colecciones.NATURALEZA): "Chal",
            (Tipo.CAMISA, Intervenido.Colecciones.GEOMETRIA): "Corte Cuadrado",
            (Tipo.CAMISA, Intervenido.Colecciones.INDUSTRIAL): "Botones",
            (Tipo.CAMISA, Intervenido.Colecciones.NATURALEZA): "Manga Sisa",
            (Tipo.PANTALON, Intervenido.Colecciones.GEOMETRIA): "Bermuda",
            (Tipo.PANTALON, Intervenido.Colecciones.INDUSTRIAL): "Traje",
            (Tipo.PANTALON, Intervenido.Colecciones.NATURALEZA): "Short",
        }

        corte = mapeo_cortes.get((tipo, self.coleccion))
        if corte:
            self.corte = corte

    def estampar(self):
        tipo = self.get_tipo()
        mapeo_estampados = {
            (Tipo.ABRIGO, Intervenido.Colecciones.GEOMETRIA): "Cuadrado",
            (Tipo.ABRIGO, Intervenido.Colecciones.INDUSTRIAL): "Manchas",
            (Tipo.ABRIGO, Intervenido.Colecciones.NATURALEZA): "Hojas",
            (Tipo.CAMISA, Intervenido.Colecciones.GEOMETRIA): "Triangulo",
            (Tipo.CAMISA, Intervenido.Colecciones.INDUSTRIAL): "Edificio",
            (Tipo.CAMISA, Intervenido.Colecciones.NATURALEZA): "Palma",
            (Tipo.PANTALON, Intervenido.Colecciones.GEOMETRIA): "Rombos",
            (Tipo.PANTALON, Intervenido.Colecciones.INDUSTRIAL): "Rayas",
            (Tipo.PANTALON, Intervenido.Colecciones.NATURALEZA): "Flores",
        }

        estampado = mapeo_estampados.get((tipo, self.coleccion))
        if estampado:
            self.estampado = estampado

    def fijacion_de_precio(self):
        costo=self.costo
        costo=int(costo)
        if self.coleccion == Intervenido.Colecciones.NATURALEZA:
            self.precio = costo + 200000
        elif self.coleccion == Intervenido.Colecciones.INDUSTRIAL:
            self.precio = costo + 300000
        elif self.coleccion == Intervenido.Colecciones.GEOMETRIA:
            self.precio = costo + 400000
        else:
            self.precio = self.get_costo()

    @staticmethod
    def paleta_de_colores(coleccion):
        colores=[]
        if coleccion == Intervenido.Colecciones.NATURALEZA:
            colores = ["VERDE", "AMARILLO", "CAFE", "BEIGE", "OLIVO", "AZUL", "NEGRO"]
        elif coleccion == Intervenido.Colecciones.GEOMETRIA:
            colores = ["AZUL", "NARANJADO", "ROJO", "VERDE", "AMARILLO", "VIOLETA", "NEGRO"]
        elif coleccion == Intervenido.Colecciones.INDUSTRIAL:
            colores = ["GRIS", "AZUL", "VIOLETA", "VERDE", "PLATEADO", "ROJO", "NEGRO"]
        return colores
    @staticmethod
    def colores(coleccion):
        colores=[]
        if coleccion == Intervenido.Colecciones.NATURALEZA:
            colores = ["#00FF00", "#FFFF00", "#964B00", "#F5F5DC", "#808000", "#ADD8E6", "#000000"]
        elif coleccion == Intervenido.Colecciones.GEOMETRIA:
            colores = ["#0000FF", "#FFA500", "#FF0000", "#00FF00", "#FFFF00", "#EE82EE", "#000000"]
        elif coleccion == Intervenido.Colecciones.INDUSTRIAL:
            colores = ["#999999", "#6666CC", "#8080A0", "#669966", "#AAAAAA", "#CC6666", "#555555"]
        return colores

    @staticmethod
    def intervenir(productos, coleccion):
        intervenidos = []
    
        camisas = Producto.clasificar(productos, Tipo.CAMISA)
        pantalones = Producto.clasificar(productos, Tipo.PANTALON)
        abrigos = Producto.clasificar(productos, Tipo.ABRIGO)
        from Python.src.gestorAplicacion.clasesHerencia.Empleado import Empleado
        CARGOS=Empleado.CARGOS
        operarios = Empleado.seleccionar_empleados(CARGOS.OPERARIO, camisas, pantalones, abrigos)
        paleta_de_colores = Intervenido.paleta_de_colores(coleccion)

        if operarios:
            for producto in productos:
                intervenido = Intervenido(producto, coleccion)
                intervenido.cortar()
                intervenido.pintar(paleta_de_colores)
                intervenido.estampar()
                intervenido.fijacion_de_precio()
                intervenidos.append(intervenido)

        return intervenidos

    # Getters
    def get_coleccion(self):
        return self.coleccion

    def get_color(self):
        return self.color

    def get_corte(self):
        return self.corte

    def get_precio(self):
        return self.precio

    def get_estampado(self):
        return self.estampado

    # Setters
    def set_coleccion(self, coleccion):
        self.coleccion = coleccion

    def set_color(self, color):
        self.color = color

    def set_corte(self, corte):
        self.corte = corte

    def set_precio(self, precio):
        self.precio = precio

    def set_estampado(self, estampado):
        self.estampado = estampado

   

    
